diag_manager/fms_diag_buffer.F90

```diff
@@ -0,0 +1,1546 @@
+  end select
+  if (allocated(this%buffer_id)) deallocate(this%buffer_id)
+  if (allocated(this%count_0d)) deallocate(this%count_0d)
+  if (allocated(this%num_elements)) deallocate(this%num_elements)
+  if (allocated(this%buffer_dims)) deallocate(this%buffer_dims)
+end subroutine flush_buffer
+
+!! -----------Type-specific routines for buffer0-5d
+
+!! allocations could be done in one routine for 0-5d if buffobj is changed to fmsDiagBuffer_class
+!! not sure which approach would be better
+
+!> allocates scalar buffer data to the given buff_type
+subroutine allocate_buffer_0d(this, buff_type, diurnal_samples, field_name)
+  class(buffer0d_type), intent(inout), target :: this !< scalar buffer object
+  class(*),intent(in) :: buff_type !< allocates to the given type, value does not matter
+  integer, optional :: diurnal_samples !< number of diurnal samples, passed in from diag_yaml
+  integer :: n_samples !< number of diurnal samples, defaults to 1
+  character(len=*), optional, intent(in) :: field_name !< optional field name for error output
+  character(len=128)          :: field_name_str
+  if(present(field_name)) then
+    field_name_str = field_name(1:128)
+  else
+    field_name_str = 'N/A'
+  endif
+
+  if(present(diurnal_samples)) then
+    n_samples = diurnal_samples
+  else
+    n_samples = 1
+  endif
+
+  if(allocated(this%buffer)) call mpp_error(FATAL, "allocate_buffer_0d: buffer already allocated for field:"// &
+                                                    field_name_str)
+  select type (buff_type)
+    type is (integer(kind=i4_kind))
+      allocate(integer(kind=i4_kind) :: this%buffer(1))
+      allocate(integer(kind=i4_kind) :: this%counter(1))
+      allocate(integer(kind=i4_kind) :: this%count_0d(n_samples))
+      this%counter = 0
+      this%buffer_type = i4
+    type is (integer(kind=i8_kind))
+      allocate(integer(kind=i8_kind) :: this%buffer(1))
+      allocate(integer(kind=i8_kind) :: this%counter(1))
+      allocate(integer(kind=i8_kind) :: this%count_0d(n_samples))
+      this%counter = 0
+      this%buffer_type = i8
+    type is (real(kind=r4_kind))
+      allocate(real(kind=r4_kind) :: this%buffer(1))
+      allocate(real(kind=r4_kind) :: this%counter(1))
+      allocate(real(kind=r4_kind) :: this%count_0d(n_samples))
+      this%counter = 0
+      this%buffer_type = r4
+    type is (real(kind=r8_kind))
+      allocate(real(kind=r8_kind) :: this%buffer(1))
+      allocate(real(kind=r8_kind) :: this%counter(1))
+      allocate(real(kind=r8_kind) :: this%count_0d(n_samples))
+      this%counter = 0
+      this%buffer_type = r8
+    class default
+       call mpp_error("allocate_buffer_0d", &
+           "The buff_type value passed to allocate a buffer is not a r8, r4, i8, or i4" // &
+           "for field:" // field_name_str, &
+           FATAL)
+  end select
+
+  allocate(this%num_elements(n_samples))
+  allocate(this%buffer_dims(1))
+  this%num_elements = 0
+  this%count_0d   = 0
```

Comment:
How are you allowed to do this if the type is assigned in the allocate statements above?  If this is a real, the `0` will also have to be implicitly converted to a real.